horovod/torch/__init__.py

```diff
@@ -72,22 +77,38 @@ def _register_hooks(self):
                     grad_acc.register_hook(self._make_hook(p))
                     self._grad_accs.append(grad_acc)
 
+    def _all_reduce_grad(self, p):
+        name = self._parameter_names.get(p)
+        tensor = p.grad.data
+        tensor_compressed, ctx = self._compression.compress(tensor)
+
+        handle = allreduce_async_(tensor_compressed, average=True, name=name)
+        return handle, ctx
+
     def _make_hook(self, p):
         def hook(*ignore):
-            assert p not in self._handles
+            if p in self._handles and self._handles[p][0] is not None:
+                if self._reduce_gradients:
+                    raise AssertionError(
+                        "Gradients were accumulated twice without a "
+                        "call to step(). Use ignore_gradients(True) to "
+                        "accumulate gradients locally.")
             assert not p.grad.requires_grad
-            name = self._parameter_names.get(p)
-
-            tensor = p.grad.data
-            tensor_compressed, ctx = self._compression.compress(tensor)
-
-            handle = allreduce_async_(tensor_compressed, average=True, name=name)
+            handle, ctx = None, None
+            if self._reduce_gradients:
+                handle, ctx = self._all_reduce_grad(p)
             self._handles[p] = (handle, ctx)
         return hook
 
     def synchronize(self):
         for p, value in self._handles.items():
             handle, ctx = value
+            if not self._reduce_gradients:
+                warnings.warn("Attempting to synchronize an optimizer that "
```

Comment:
We should mention why it's bad (by default gradients are allreduced asynchronously).  Not sure we should fallback / mention it, since it may create an illusion of "no problem".

Reply:
With new `backward_passes_per_step`, we can say something like `Called step on optimizer before computing gradients backwards_passes_per_step times.  This will cause performance degradation, as gradient exchange is interleaved with computation only on the backwards_passes_per_step'th pass.`